python/pyiceberg/io/pyarrow.py

```diff
@@ -23,10 +23,16 @@
 """
 
 import os
-from functools import lru_cache
-from typing import Callable, Tuple, Union
+from functools import lru_cache, singledispatch
+from typing import (
+    Callable,
+    List,
+    Tuple,
+    Union,
+)
 from urllib.parse import urlparse
 
+import pyarrow as pa
```

Comment:
Is this file the right location for `pyarrow` conversions? This is an implementation of `FileIO` so I don't know that it fits to put schema conversion here.

Reply:
I was also not sure, but it felt like me the right place mainly because PyArrow is optional, and if so, we only need to be careful not to import this Python module. If we would put it in `schema.py`, then we need to have logic to make sure that it is optional, and if you don't have pyarrow installed, everything still works.

Reply:
Maybe we will want to move this to an `extensions` module eventually. It is weird to call into `pyiceberg.io` for this functionality.